kv/kvnemesis: TestValidate failed

kv/kvnemesis.TestValidate [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13135887?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/13135887?buildTab=artifacts#/) on master @ [cbcdca3b6a5e54ee936efe17c854a73bbbab74af](https://github.com/cockroachdb/cockroach/commits/cbcdca3b6a5e54ee936efe17c854a73bbbab74af):


```
=== RUN   TestValidate
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/1f42cf5be2fc021646bf9b2daf5eaef3/logTestValidate3099122501
    test_log_scope.go:81: use -show-logs to present logs inline
--- FAIL: TestValidate (0.65s)
=== RUN   TestValidate/single_mvcc_rangedel
    --- FAIL: TestValidate/single_mvcc_rangedel (0.00s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
 - <code>stress=true</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestValidate.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-34626

```
    --- FAIL: TestValidate/single_mvcc_rangedel (0.00s)
panic: RangeKeySet called with suffixed start key [recovered]
	panic: RangeKeySet called with suffixed start key

goroutine 4478 [running]:
testing.tRunner.func1.2({0x767b460, 0xbf63910})
	GOROOT/src/testing/testing.go:1545 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1548 +0x716
panic({0x767b460?, 0xbf63910?})
	GOROOT/src/runtime/panic.go:920 +0x270
github.com/cockroachdb/pebble.(*Batch).RangeKeySet(0xc0059c2d80, {0xc005958978, 0x14, 0x18}, {0xc005958990, 0x14, 0x18}, {0xc004431587, 0x9, 0x9}, ...)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/batch.go:938 +0x4b8
github.com/cockroachdb/pebble.(*DB).RangeKeySet(0xc005e79b00, {0xc005958978, 0x14, 0x18}, {0xc005958990, 0x14, 0x18}, {0xc004431587, 0x9, 0x9}, ...)
	github.com/cockroachdb/pebble/external/com_github_cockroachdb_pebble/db.go:734 +0x165
github.com/cockroachdb/cockroach/pkg/kv/kvnemesis.(*Engine).DeleteRange(0xc002921e20, {0xc005958978, 0x14, 0x18}, {0xc005958990, 0x14, 0x18}, {0x1, 0x0, 0x0}, ...)
	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/engine.go:122 +0x133
github.com/cockroachdb/cockroach/pkg/kv/kvnemesis.TestValidate.func9(0xc00529eb60?)
	github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/pkg/kv/kvnemesis/validator_test.go:2238 +0x84b
github.com/cockroachdb/cockroach/pkg/testutils/echotest.(*Walker).Run.func2(0xc00529eb60)
	github.com/cockroachdb/cockroach/pkg/testutils/echotest/echotest.go:170 +0x54
testing.tRunner(0xc00529eb60, 0xc002921800)
	GOROOT/src/testing/testing.go:1595 +0x262
created by testing.(*T).Run in goroutine 3011
	GOROOT/src/testing/testing.go:1648 +0x846
```

--- a/pkg/kv/kvnemesis/engine.go
+++ b/pkg/kv/kvnemesis/engine.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/storage"
 	"github.com/cockroachdb/cockroach/pkg/util/bufalloc"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/pebble"
 	"github.com/cockroachdb/pebble/vfs"
 )
@@ -119,7 +120,9 @@ func (e *Engine) Put(key storage.MVCCKey, value []byte) {
 
 func (e *Engine) DeleteRange(from, to roachpb.Key, ts hlc.Timestamp, val []byte) {
 	suffix := storage.EncodeMVCCTimestampSuffix(ts)
-	if err := e.kvs.RangeKeySet(from, to, suffix, val, nil); err != nil {
+	err := e.kvs.RangeKeySet(
+		storage.EngineKey{Key: from}.Encode(), storage.EngineKey{Key: to}.Encode(), suffix, val, nil)
+	if err != nil {
 		panic(err)
 	}
 }
@@ -153,18 +156,27 @@ func (e *Engine) Iterate(
 			}
 		}
 		if iter.RangeKeyChanged() {
-			key, endKey := iter.RangeBounds()
-			e.b, key = e.b.Copy(key, 0 /* extraCap */)
-			e.b, endKey = e.b.Copy(endKey, 0 /* extraCap */)
+			keyCopy, endKeyCopy := iter.RangeBounds()
+			e.b, keyCopy = e.b.Copy(keyCopy, 0 /* extraCap */)
+			e.b, endKeyCopy = e.b.Copy(endKeyCopy, 0 /* extraCap */)
 			for _, rk := range iter.RangeKeys() {
 				ts, err := storage.DecodeMVCCTimestampSuffix(rk.Suffix)
 				if err != nil {
 					fn(nil, nil, hlc.Timestamp{}, nil, err)
 					continue
 				}
+				engineKey, ok := storage.DecodeEngineKey(keyCopy)
+				if !ok || len(engineKey.Version) > 0 {
+					fn(nil, nil, hlc.Timestamp{}, nil, errors.Errorf("invalid key %q", keyCopy))
+				}
+				engineEndKey, ok := storage.DecodeEngineKey(endKeyCopy)
+				if !ok || len(engineEndKey.Version) > 0 {
+					fn(nil, nil, hlc.Timestamp{}, nil, errors.Errorf("invalid key %q", endKeyCopy))
+				}
 
 				e.b, rk.Value = e.b.Copy(rk.Value, 0)
-				fn(key, endKey, ts, rk.Value, nil)
+
+				fn(engineKey.Key, engineEndKey.Key, ts, rk.Value, nil)
 			}
 		}
 	}

--- a/pkg/kv/kvnemesis/validator.go
+++ b/pkg/kv/kvnemesis/validator.go
@@ -1058,8 +1058,10 @@ func (v *validator) checkAtomicCommitted(
 					panic(err)
 				}
 			} else { // ranged write
+				key := storage.EngineKey{Key: o.Key}.Encode()
+				endKey := storage.EngineKey{Key: o.EndKey}.Encode()
 				suffix := storage.EncodeMVCCTimestampSuffix(o.Timestamp)
-				if err := batch.RangeKeyUnset(o.Key, o.EndKey, suffix, nil); err != nil {
+				if err := batch.RangeKeyUnset(key, endKey, suffix, nil); err != nil {
 					panic(err)
 				}
 			}
@@ -1103,8 +1105,10 @@ func (v *validator) checkAtomicCommitted(
 					panic(err)
 				}
 			} else {
+				key := storage.EngineKey{Key: o.Key}.Encode()
+				endKey := storage.EngineKey{Key: o.EndKey}.Encode()
 				suffix := storage.EncodeMVCCTimestampSuffix(writeTS)
-				if err := batch.RangeKeySet(o.Key, o.EndKey, suffix, o.Value.RawBytes, nil); err != nil {
+				if err := batch.RangeKeySet(key, endKey, suffix, o.Value.RawBytes, nil); err != nil {
 					panic(err)
 				}
 			}

